### DISCUSSION TITLE
Proactive reminders

### DISCUSSION THREAD

**Discussion by dstrelnikov** [iSoron/uhabits#2045] [Feature Requests]
I believe that the frequency of the reminders should be decoupled from the target frequency of the habit. I suggest the following concept of **proactive** reminders.

## Motivation

Let us consider two following habit models.

1. **Habit A** delivers benefit when executed as often as possible, ideally every day. However, if it is not executed every day it is not a big deal and the habit still delivers a lot of benefits. Moreover, due to the nature of the habit it could be that the life circumstances just do not allow for the daily execution.

Therefore, to not get obsessed with it, to be realistic, and to track the real impact of the habit on my life the target frequency could be set to something like "2 times every 3 days".

I would like though to get a reminder for **Habit A** every day to look for a possibility to execute it and make my own decision whether it should be done today.

2. In contrast to the previous example, **Habit B** must be executed once a week and ideally no more often. One does not need to see the reminder for **Habit B** every day. In real life however strict execution of **Habit B** every 7 days is sometimes not possible. In such rare cases I would like to be proactive and execute the habit after only 6 days.

In order to look for a possibility to execute it and to decide whether it should be after 6 or 7 days, I would like to start getting a reminder for **Habit B** one day in advance before its real due date.

## Concept of reminder proactivity

At the moment the reminders are coupled with the habit target frequency and are reactive, i.e. the reminder will only appear on the next "uncovered" day. In real life with some habits this leads to missing the possibility to execute them in time, leaving gaps in the tracker.

I suggest to add an option to make the reminders **proactive**, e.g. by letting the user to specify how many days in advance the reminder should start appearing. This could be zero, i.e. the current behavior, or `inf` (or any number greater than the maximal possible gap withing the set target frequency) for getting a reminder every day, as well as anything in between.

## Why making things more complicated?

I'm a big advocate of keeping things as simple as possible. One could argue that proactive reminders are not necessary since one could achieve similar results by adjusting the workflow utilizing the current app functionality.

For instance one could set the target frequency for **Habit A** to be "every day" and for **Habit B** to be "1 times per 6 days" in order to get the desired reminders and simply accept the gaps in the tracker as a part of imperfect real life.

I would argue that the gamification and "positive feedback" aspects of the habit tracker have a great positive impact on individual's mental health and improve adherence. It is so satisfying to explicitly or implicitly check all the habit every single day and see that you are doing "the right thing" withing the framework of your goals and the habits introduced to help achieving them.

Leaving gaps in the tracker in my opinion does not allow for the same satisfying feeling even if you know that you did everything right. A habit tracker is an important tool that should reflect as precisely as possible whether you are on track.

Another reason against accepting gaps in the tracker is the following. Once your brain learns that some gaps are okay because you did your best and it's just a technical limitation of the software, more gaps can start slipping unnoticed, see the [broken windows theory](https://en.wikipedia.org/wiki/Broken_windows_theory).

In my opinion the proper use of the habit tracker is to aim that every single day every single habit is explicitly or implicitly checked. Gaps in the tracker should be a signal to revise the system, to diagnose what exactly is happening and why.

## Related

- https://github.com/iSoron/uhabits/discussions/686

**Comment by BURG3R5** [comment]
I'm interested in a solution for the **Habit A** model as well.

We could add an option to have a separate "reminder frequency", which will be used to determine when to send reminders instead of the habit's frequency. So the habit can be once a week while the reminder is sent once every two days. Maybe the notification can include an indicator whether or not the habit is already implicitly completed.

@iSoron If this idea (and this solution) seems okay, I can start working on it.

**Reply by dstrelnikov** [reply]
Hey @BURG3R5! Thanks for picking up on this!

I just wanted to clarify a bit my concept here. Both Habit A and Habit B workflows can be implemented with the same "reminder proactivity [days]" setting. This setting would determine the number of days (in advance) when the reminder starts showing before the habit is "due" (i.e. when the reminder appears in the current version). If the "proactivity" is set to `2 days` then it will give you a 2 day margin to act on the habit before it's due. If it's set to `0 days` then you will be reminded every day. You see the idea? I was thinking about it for a while and I believe this would be the most elegant approach to cover different workflows.

**Reply by BURG3R5** [reply]
I've researched a bit about my solution and found that it will probably cause issues with backwards-compatibility. And your solution _does_ fit my personal use-case (without being a breaking change), so I'm up for implementing that too.

**Comment by KyleSCraig** [comment]
For Habit B I backdate execution. So if I normally do it on Saturday but sometimes Friday or Sunday are better, I mark it as Friday regardless. You have to open the app to do this.
